packages/data-broker/src/__tests__/services/dhis/pullers/EventsPuller.test.js

```diff
@@ -106,20 +102,25 @@ describe('EventsPuller', () => {
             },
             dimensions: { POP01: [], POP02: [] },
           },
-          rows: [],
+          rows: [
+            [5, 8],
+            [4, 3],
+          ],
         };
         dhisApi = createMockDhisApi({ getEventAnalyticsResponse });
         stubGetDhisApi(dhisApi);
         const dataElementCodes = ['POP01', 'POP02'];
+        const organisationUnitCodes = ['DL_FACILITY_A'];
 
-        await eventsPuller.pull([dhisApi], [DATA_GROUPS.POP01_GROUP], {
+        const events = await eventsPuller.pull([dhisApi], [DATA_GROUPS.POP01_GROUP], {
           dataElementCodes,
+          organisationUnitCodes,
         });
-        expect(buildEventsFromDhisEventAnalytics).toHaveBeenCalledOnceWith(
-          expect.anything(),
-          getEventAnalyticsResponse,
-          dataElementCodes,
-        );
+
+        expect(events).toEqual([
+          { dataValues: { POP01: 5, POP02: 8 } },
+          { dataValues: { POP01: 4, POP02: 3 } },
+        ]);
```

Comment:
Reworked these tests to actually return real data. I always prefer that over testing spy implementations where possible 👍

Reply:
Ah yep this is the `buildEventsFromDhisEventAnalytics() invocation` describe block, that's why that's mocked out like that. It's not testing real data because it's more of a unit test with dependencies mocked out, so that changes to buildEventsFromDhisEventAnalytics don't fail this test, and so that buildEventsFromDhisEventAnalytics is not double-tested. 

I like proper full integration tests too, but my feeling is this area is so complicated we need to break things up if only for reducing cognitive load when working in this part of the codebase, so unit tests are great for this complexity management and processes like TDD, when you are working on a single class you can work with its tests alone and focus on what it alone does and not the half dozen or so classes it's connected to.

We do definitely need proper integration tests as well though, especially for DhisService, because it's so complicated, the rest are probably fine. Something like DhisService.integration.test.js would be a good place, following what we do in the indicators package with src/__tests__/__integration/... Otherwise for something simpler maybe a new describe block at the top of this file called something like 'integration'?

Reply:
Ah I see your point. And yup, my bad for changing these without taking a look at how we've structured tests in this package.

I'm thinking of creating a separate describe block within this `EventsPuller` test file. I guess the rationale for doing so would be that ideally the logic around handing tracked entity instances is hidden within `EventsPuller`. I wouldn't necessarily want the tests to need to go in an integration test suite as it's a bit of an implementation detail. 

I'll restore the old tests to their previous behaviour 👍 

Also, fwiw I've never been too concerned about double testing areas of code. It's always felt more effort than its worth to maintain trying to just have a single test to cover each code path? Thoughts?

Reply:
Yeah I was thinking about where best to place those, a block in EventsPuller makes sense because it's pretty specific to that. The other options of DhisService is nice for being able to see all possible input/outputs of DhisService at a glance, but we typically only care about analytics or events not both at the same time. Then there's the option of top level / DataBroker level, which is also nice, for seeing and testing all services, would be cool being able to see which services support what, and it would be a real integration test going through all layers, but again we're not jumping between them very often so it's not super beneficial day to day. I like your idea of placement.

Double testing, yeah it's a judgement call. I'm pretty sure none of our unit tests are actually pure unit tests, there's always something we don't mock out. But the value I see of unit tests and TDD in general is that when a test is nice and neat and tests one thing it means your class/function must be neat and only do that one thing. So it kind of forces separation of concerns and clean code. You can have a function X that does two things and a test Y that tests that it does both things, but you can't have a test Y that tests one thing and a function X that does two things (if that makes sense).

Reply:
Yeah definitely agree that a high level place where we have a suite of tests that show the basic api (input -> output) would be really nice. This is a notoriously opaque interface imo.

But I'm always a fan of having tests that get into the details being around the code that handles those details.

>You can have a function X that does two things and a test Y that tests that it does both things, but you can't have a test Y that tests one thing and a function X that does two things (if that makes sense). I tend to think of it in those terms, rather than avoiding two tests overlapping.

That's a really good way of describing it! Thanks for sharing that perspective!

I suspect I lean towards the idea that when logic is composed hierarchically (eg. EventsPuller calls EventsTranslator, and EventsBuilder), that it seems reasonable that it should test both of those during its tests. Not explicitly, but just as a natural side effect of writing simple tests that are just aware of the interface they're testing (and not the underlying behaviour). 
Of course mocking is required at the service boundary layer (database, external package/api) but other than that I prefer to keep it to a minimum. I have a bit of an anti-mocking bias tho, which admittedly comes from some valid places but also just a bit of discomfort with the mechanics of it which isn't really a valid concern 😅

Reply:
Mocking in javascript is so painful 😢 It really is. I understand the bias completely 😆 

>writing simple tests that are just aware of the interface they're testing (and not the underlying behaviour).

Yes! This is something javascript doesn't do well, there's no inversion of control / dependency injection which would let us do this. I think having our tests/mocking know about the implementation is just a pain point we have to bear :/ It seems to be the way jest and them recommend so we're not alone.

>I suspect I lean towards the idea that when logic is composed hierarchically (eg. EventsPuller calls EventsTranslator, and EventsBuilder), that it seems reasonable that it should test both of those during its tests.

I agree for sure. With this one you definitely benefit from integration testing the parent with its children, you make sure these three parts are working well together, and like you said above, you get simplicity and not having to do this weird mocking-based-on-implementation-knowledge thing.

The downside I see is when it gets too complex to do this. You are super familiar about this codebase and a really good dev so it's no biggie for you keeping all this in your head :) but ask a junior to make a change in data-broker and their head would explode... I mean I've spent two years on this and my head is still exploding... Anyway treating the children and parent as one things just means the junior would need to understand everything at the same time, i.e. the full implementation to make a change. The TDD way would still ask a lot from them, but once they know just the point of each thing, or the single responsibility of each class, skim through the tests maybe, they can then more easily figure out where and how to make a change.

So yes I agree with you, except in this one specific part of the codebase where my head hurts thinking about everything at the same time :)